File: src/main/kotlin/icu/takeneko/libNeko/nyatwork/Writer.java
package icu.takeneko.libNeko.nyatwork;

import java.util.Optional;
import java.util.function.BiConsumer;

@FunctionalInterface
public interface Writer<T> extends BiConsumer<FriendlyByteBuf, T> {
    default Writer<Optional<T>> asOptional() {
        return (buf, optional) -> {
            buf.writeOptional(optional, this);
        };
    }
}

File: src/main/kotlin/icu/takeneko/libNeko/nyatwork/Writer.kt
package icu.takeneko.libNeko.nyatwork

import java.util.*
import java.util.function.BiConsumer

fun interface Writer<T: Any> : BiConsumer<FriendlyByteBuf, T> {
    fun asOptional(): Writer<Optional<T>> {
        return Writer { buf: FriendlyByteBuf, optional: Optional<T> ->
            buf.writeOptional(optional, this)
        }
    }
}

